### PR TITLE
insights: add historical data enqueuer

### LINKED ISSUE #18398: insights: backfill data

### PR DESCRIPTION
**Pull request by slimsag**
Here's the deal: This PR makes backend insights useful. Without this, it takes months for insights to produce any data - effectively making it useless.

The first thing you should do is read this high-level explanation of what this does:

https://github.com/sourcegraph/sourcegraph/blob/6b3438f5574391f8bbedde6b9708d59eca827209/enterprise/internal/insights/background/historical_enqueuer.go#L29-L52

The second thing you should do is read this high-level explanation of what the code is doing:

https://github.com/sourcegraph/sourcegraph/blob/6b3438f5574391f8bbedde6b9708d59eca827209/enterprise/internal/insights/background/historical_enqueuer.go#L113-L145

Additional context: I've spent almost 2 weeks now on this code/problem, and this was the best solution I could come up with. I did a lot (see [these ~6 PRs](https://github.com/sourcegraph/sourcegraph/pulls?q=is%3Apr+label%3Ainsights-historical+is%3Aclosed)) already to try and reduce the problem to its individual components as much as possible, but it's still complex in nature. The code is good, the tests are good, but it's complex. More complex than I'd like. I would love to reduce complexity here further, but I don't have more time to spend on this - I will have to leave it up to the next person. I'm open to quick ways I can improve clarity.

Note: On a dev server this takes around 10-20 minutes to backfill data for two search insights. On larger instances with many more repositories, it is expected to take hours. I will begin to test throughput and load on gitserver/search/repo-updater in order to tune this further after merging. We benefit from people scaling up gitserver, but do not want to _cause_ anyone to have to do so.

Fixes #18398

Signed-off-by: Stephen Gutekanst <stephen@sourcegraph.com>

### CODE REVIEW

File: enterprise/internal/insights/background/historical_enqueuer.go [hunk 1]
```diff
@@ -0,0 +1,454 @@
...
+//
+// 1. Cache duplicative calls to prevent performing heavy operations multiple times.
+// 2. Lift heavy operations to the layer/loop one level higher, when it is sane to do so.
+// 3. Ensure we perform work slowly, linearly, and with yielding/sleeping between any substantial
+//    work being performed.
+//
+type historicalEnqueuer struct {
+	// Required fields used for mocking in tests.
+	now                   func() time.Time
+	sleep                 func(t time.Duration)
```
Comment (efritz) [inline enterprise/internal/insights/background/historical_enqueuer.go:148]: Is this easier than using glock?

**Review by efritz — APPROVED** [review]
I'm going to play around with this a bit to see if I can give you any insights on how to better get away from the `yield()` trickery here.

Nothing raises alarms here, but I'd like to know your thoughts about multiple frontends all enqueueing jobs. Will they double-queue work as it's currently written (I couldn't tell at a glance), or will you need to add some sort of advisory locking mechanism around the enqueue?

File: enterprise/internal/insights/background/historical_enqueuer.go (hunk 1 above)
Comment (slimsag) [reply enterprise/internal/insights/background/historical_enqueuer.go:148]: I didn't know glock was a thing, but it seems about equal so I'll leave as-is for now and refactor to use glock later

**Comment by slimsag** [comment]
> I'd like to know your thoughts about multiple frontends all enqueueing jobs. Will they double-queue work as it's currently written (I couldn't tell at a glance), or will you need to add some sort of advisory locking mechanism around the enqueue?

Frontends don't enqueue jobs, this is in repo-updater singleton where the rest of the insights background jobs run.

**Pull request merged**